tokio/src/sync/semaphore.rs

```diff
@@ -47,28 +47,44 @@ use std::sync::Arc;
 /// }
 /// ```
 ///
-/// Use [`Semaphore::acquire_owned`] to move permits across tasks:
+/// Limit number of incoming requests being handled at the same time.
 ///
-/// ```
+/// Similar to limiting the numer of simultaneous opened files, network handles are a limited resource.
+/// Allowing an unbounded amount of requests to be processed could result in a denial-of-service,
+/// and many other issues.
+///
+/// However, in contrast to the file example, this example uses a non-static, `Arc`-wrapped `Semaphore`
+/// for more fine-grained lifetime and ownership control. The `Arc` allows multiple threads and tasks
+/// to share ownership of the semaphore. Here we create a shallow-copy of the `Arc<Semaphore>` reference
+/// using `clone`. Then we acquire a permit through the `Arc` from the `Semaphore` via [`Semaphore::acquire_owned`],
+/// and move it inside the task. This ensures no non-`'static` variables are referenced from within
+/// said task.
+///
+/// If we leave the scope where the `Arc<Semaphore>` was defined, and references still exist in any
+/// of our tasks to it, the `Semaphore` will continue to exist in a static-like context, until all
+/// `Arc`s have been dropped.
+///
+/// ```compile_fail
 /// use std::sync::Arc;
 /// use tokio::sync::Semaphore;
+/// use tokio::net::{TcpListener};
 ///
 /// #[tokio::main]
-/// async fn main() {
+/// async fn main() -> Result<(), Box<dyn std::error::Error>> {
 ///     let semaphore = Arc::new(Semaphore::new(3));
-///     let mut join_handles = Vec::new();
+///     let listener = TcpListener::bind("127.0.0.1:8080").await?;
 ///
-///     for _ in 0..5 {
+///     loop {
+///         // Acquire permit before accepting, to avoid needlessly taking up a (finite) file descriptor
 ///         let permit = semaphore.clone().acquire_owned().await.unwrap();
-///         join_handles.push(tokio::spawn(async move {
-///             // perform task...
-///             // explicitly own `permit` in the task
-///             drop(permit);
-///         }));
-///     }
+///         let (mut socket, _) = listener.accept().await?;
 ///
-///     for handle in join_handles {
-///         handle.await.unwrap();
+///         tokio::spawn(async move {
+///             // Do work, and drop socket after
+///             handle_connection(&mut socket).await;
+///             // Explicitly drop the permit, so more tasks can be created
+///             drop(permit);
```

Comment:
```suggestion
///             // Do work using the socket.
///             handle_connection(&mut socket).await;
///             // Drop socket while the permit is still live.
///             drop(socket);
///             // Drop the permit, so more tasks can be created.
///             drop(permit);
```

Reply:
I'll add it, but I was going to ask about this: if we move the reference in, wouldn't that not automatically drop the socket? 

For example, if we were going for an implicit `drop(socket)` at the end of our `handle_connection` wouldn't the right way to go about it be just `mut socket`?

Reply:
You're not wrong. Taking ownership of `socket` would probably be nicer here, if this was real code. My motivation is that I think the example will be more helpful this way.

When people write this kind of code themselves, they will either:

 * Move the socket into another function.
 * Not move the socket into another function.

If they're basing their code on this example and moving the socket, then I'm sure that they will figure out that they don't need the `drop(socket)`. However, if they're not moving it into another function (e.g., maybe they process it directly in `tokio::spawn` like in the readme), then there's a good chance that they forget the `drop(socket)`, and I want the example to remind them.